src/nodes/chunk_dispatch/chunk_insert_state.c

```diff
@@ -572,17 +573,27 @@ ts_chunk_insert_state_create(const Chunk *chunk, ChunkDispatch *dispatch)
 													  ALLOCSET_DEFAULT_SIZES);
 	OnConflictAction onconflict_action = chunk_dispatch_get_on_conflict_action(dispatch);
 	ResultRelInfo *relinfo;
+	const Chunk *chunk;
 
 	/* permissions NOT checked here; were checked at hypertable level */
-	if (check_enable_rls(chunk->table_id, InvalidOid, false) == RLS_ENABLED)
+	if (check_enable_rls(chunk_relid, InvalidOid, false) == RLS_ENABLED)
 		ereport(ERROR,
 				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 				 errmsg("hypertables do not support row-level security")));
-	Assert(chunk->relkind == RELKIND_RELATION || chunk->relkind == RELKIND_FOREIGN_TABLE);
 
-	ts_chunk_validate_chunk_status_for_operation(chunk, CHUNK_INSERT, true);
+	DEBUG_WAITPOINT("chunk_insert_before_lock");
+	rel = table_open(chunk_relid, RowExclusiveLock);
```

Comment:
Is my understanding correct that the compression code [locks](https://github.com/timescale/timescaledb/blob/8875c7a8970e9d60d386f0524b63f6e54025397e/tsl/src/compression/api.c#L462) the source chunk with an `ExclusiveLock` (which conflicts with our `RowExclusiveLock`). So, the chunk meta data cannot be changed after we have obtained this lock?

Maybe you could add a comment to the code to make this behavior more clear.

Reply:
This isn't about lock level. The PR doesn't change the locks taken only the place where it is taken.

The issue is that you cannot trust that metadata is up-to-date unless you read it after taking the lock since the chunk could be altered between reading the metadata and taking the lock.

There are many operations that can alter metadata, so compression/decompression is only one such case that happened to trigger the bug.

I can add a comment about the specific case of concurrency between compression and insert, but this is actually more general than that.

Reply:
Thanks for the explanation and sorry for the confusion. It seems my question was not precisely formulated. I was wondering if another process could change the metadata even _after_ taking the lock in line 585 and reading the metadata. 

`RowExclusiveLock` is one of the weaker locks and it is not self-conflicting. However, the compression code takes a stronger `ExclusiveLock`, which conflicts with our `RowExclusiveLock`. So, a parallel running compression job should block and the metadata will not be modified by this job in parallel. Therefore, the change seems to be correct for this case.

However, if another process takes a weaker, non-conflicting lock, like `RowExclusiveLock,` and modifies the metadata, I am wondering if we could still face the same problem or if there is something else that prevents these situations? If we just rely on the fact that all other code paths that modify the metadata have to conflict with our `RowExclusiveLock`, we should document this assumption.

Reply:
@jnidzwetzki than you for the clarification. 

I added an extra note where the `RowExclusiveLock` is taken. Is this enough documentation for now?

Just to be clear though: I haven't changed any locks or lock-levels in this PR and conflicts were already happening.

The change I've made is about how metadata is used _around_ locks: The PR only ensures we avoid using metadata  that was read _prior_ to locking and then consistently use the up-to-date metadata  _after_ locking. There was already a comment in the code about the locks taken and how it aimed to conflict with compression which I tweaked a bit.


Also, it might be worth noting that, while re-compression starts with taking an `ExclusiveLock`,  I believe it will at some point take an `AccessExclusiveLock` to drop the relation and create a new one. This conflicts with every other lock, including `AccessShareLock`.

Thus, a process that doesn't change metadata should be able to get away with an `AccessShareLock`. The INSERT process doesn't change metadata, but takes a `RowExclusiveLock` because it modifies data (not metadata) in the relation.

If we have code that modifies a releation's metadata with a `RowExclusiveLock` or lower it would be a bug I guess. I believe this is part of the principles of PG itself, if you read the section on "Explicit Locking" where "higher" locks are always necessary to do "maintenance" type things on relations (i.e., change metadata, except for maybe stats). 

In the end, it is up to our code to "do the right thing" according to established principles around the use of locks. Unfortunately, we currently don't do this in a clean way as the code has evolved organically by many people. It is probably time for a major refactor where we also establish stricter principles and better interfaces. But this is out-of-scope of this fix.

Reply:
@erimatnor Thanks for the explanation.